apps/node-fastify/src/crons/refresh-stale-subscriptions.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Orb currently has no webhook whenever a billing cycle is reset, so subscription billing cycle information may be outdated.
+ *
+ * As a temporary solution, we use a CRON to look at active subscriptions with a billing cycle end date in the past.
+ *
+ * Orb provides an API that takes multiple customer ids when looking for subscriptions.
+ */
+
+import { OrbSync } from 'orb-sync-lib';
+import { logger } from '../utils/logger';
+
+export async function refreshStaleSubscriptions(orbSync: OrbSync) {
+  logger.info(`Starting to refresh stale subscriptions`);
+
+  const customerIdsWithOutdatedSubscriptionsData = await orbSync.postgresClient
+    .selectMany<{ customer_id: string }>(
+      `
+        SELECT customer_id FROM orb.subscriptions
+        WHERE status = 'active'
+        AND current_billing_period_end_date < :now
+        LIMIT 2500
```

Comment:
Given it runs hourly, we may have more than 2500 subscriptions per day and it may never catch up, maybe we need to set it a bit higher here

Reply:
Right now the job runs on every 15th minute. I'm not stating the exact number of our active subscriptions here, but even taking into account that the billing cycle end dates are not spread evenly across the days of a month that would be enough.
Fetching 100 subscriptions from the Orb-Api easily takes 20 something seconds. So a job would run for at least 8 minutes. Which should be fine given that the jobs run every 15 minutes.

But there's a another problem. We run the sync engine with more than one task. Even with the tasks starting up at a different point in time, the cron jobs would all start at e.g. 13:15pm, 13:30pm and thereby fetching the same "stale" subscriptions from the database. Even if it was possible to schedule the jobs for "n minutes after the last run" it wouldn't be ideal given that the Orb-Api calls take so long.

Reply:
Let's introduce a simple lock table in the orb schema then and do a simple distributed lock

Reply:
Alternatively we could look into a different approach like listening on the invoice issued webhook and getting the information from there

Reply:
Based on the plan invoice and the period of the plan line item, we should be able to "guess" the correct date

Reply:
That would also avoid the whole complexity of the cron and distributed lock

Reply:
OK, I'll look into the approach with the invoice issued webhook.